### PR TITLE
Fix NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE does not consider null val…

### PR DESCRIPTION
**Pull request by mdouaihy**
This fixes the issue reported in 

- https://github.com/spotbugs/spotbugs/issues/651
- https://github.com/spotbugs/spotbugs/issues/456

NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE is raised when using Objects.requireNonNull whereas it's not raised when using Guava Preconditions.checkNotNull.

----
Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Review by hazendaz — APPROVED** [review]

**Review by ThrawnCA — APPROVED** [review]

**Comment by gtoison** [comment]
Thanks a lot for the contribution @mdouaihy 
#651 mentionned other methods than `Objects.requireNonNull`, did you consider adding things like:
```
com.google.common.base.Verify.verifyNotNull
com.google.common.base.Preconditions.checkNotNull
org.apache.commons.lang.Validate.notNull
```

I think it would be good adding unit tests for these too

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/IssueRequireNotNullTest.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Unit test to reproduce <a href="https://github.com/spotbugs/spotbugs/issues/456">#259</a>.
+ */
+@ExtendWith(SpotBugsExtension.class)
+class IssueRequireNotNullTest {
+
+    @Test
+    void testIssue(SpotBugsRunner spotbugs) {
+        BugCollection bugCollection = spotbugs.performAnalysis(
+                Paths.get("../spotbugsTestCases/build/classes/java/main/nullnessAnnotations/TestNonNull7.class"));
+        assertThat(bugCollection, emptyIterable());
```
Comment (JuditKnoll) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/nullness/IssueRequireNotNullTest.java:26]: It's better to add a `BugInstanceMatcher` matching to the bugtype when creating negative tests, so later if a new bug will be added, which finds a hit in the testcase, then this test won't fail. See https://github.com/spotbugs/spotbugs/blob/master/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/Issue2147Test.java as an example.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java [hunk 2]
```diff
@@ -1386,7 +1386,7 @@ boolean callToAssertionMethod(Location loc) {
             if (i instanceof InvokeInstruction) {
                 InvokeInstruction ii = (InvokeInstruction) i;
                 String name = ii.getMethodName(classContext.getConstantPoolGen());
-                if (name.startsWith("check") || name.startsWith("assert")) {
+                if (name.equals("requireNonNull") || name.startsWith("check") || name.startsWith("assert")) {
```
Comment (JuditKnoll) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDeref.java:1389]: Can you please check class of the called function as well?

File: spotbugsTestCases/src/java/nullnessAnnotations/TestNonNull7.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package nullnessAnnotations;
+
+import java.io.File;
+import java.util.Objects;
+
+
```
Comment (JuditKnoll) [inline spotbugsTestCases/src/java/nullnessAnnotations/TestNonNull7.java:6]: It's a good practice to refer the issues at the testcases, if there are any.

File: CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,7 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 - Eclipse: Do not export javax.annotation packages ([[#2699](https://github.com/spotbugs/spotbugs/pull/2699)])
 - Fixed not thread safe FindOverridableMethodCall detector ([[#2701](https://github.com/spotbugs/spotbugs/issues/2701)])
 - Fix the weird messages of PI_DO_NOT_REUSE_PUBLIC_IDENTIFIERS bugs. ([#2646]((https://github.com/spotbugs/spotbugs/issues/2646)]))
+- Fix false positive NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE even if Objects.requireNonNull is used. ([[#651](https://github.com/spotbugs/spotbugs/issues/651)], [[#456](https://github.com/spotbugs/spotbugs/issues/456)])
```
Comment (JuditKnoll) [inline CHANGELOG.md:16]: Can you please remove the unnecessary brackets from the links. Now, it looks like this: ([[#651](https://github.com/spotbugs/spotbugs/issues/651)], [[#456](https://github.com/spotbugs/spotbugs/issues/456)])

**Review by JuditKnoll — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (hazendaz) [reply CHANGELOG.md:18]: oh I see, was a problem but @iloveeclipse fixed it already.  We just need this one fixed still.

**Comment by hazendaz** [comment]
Leaving this one out of 4.8.2 release as a few other items need looked at here still.

**Comment by robinjhector** [comment]
Hi, this would be great to get merged and released! Any chance you can take a look at the requested changes @mdouaihy ?

**Comment by mdouaihy** [comment]
hi @robinjhector, 

I am looking into applying the requested changes. I am just having a ClassNotFound when adding the tests to cover the changes and I am trying to figure out why.

**Comment by JuditKnoll** [comment]
> hi @robinjhector,
> 
> I am looking into applying the requested changes. I am just having a ClassNotFound when adding the tests to cover the changes and I am trying to figure out why.

In this case, it can be a good idea to convert the PR to draft.
Is the class file for not found class added to the `performAnalysis()` function inside the testclass? If yes and you are stuck, feel free to commit your changes, and I can look into it.

**Comment by hazendaz** [comment]
@mdouaihy Can you look at last concern here?   Would like to include this in next release in 2 weeks.

**Comment by mdouaihy** [comment]
@hazendaz, due to some personal problems, I was not able to fix the issues before. I'll fix that during the coming couple of days.

**Comment by hazendaz** [comment]
@JuditKnoll Any chance you could pick this up and do the necessary items you were wanting adjusted?  Seems @mdouaihy Has some personal issues to deal with so not available to make this go.  If its not too difficult, we could get this included in upcoming release hopefully this weekend :)

**Comment by JuditKnoll** [comment]
> Thanks a lot for the contribution @mdouaihy #651 mentionned other methods than `Objects.requireNonNull`, did you consider adding things like:
> 
> ```
> com.google.common.base.Verify.verifyNotNull
> com.google.common.base.Preconditions.checkNotNull
> org.apache.commons.lang.Validate.notNull
> ```
> 
> I think it would be good adding unit tests for these too



> Thanks a lot for the contribution @mdouaihy #651 mentionned other methods than `Objects.requireNonNull`, did you consider adding things like:
> 
> ```
> com.google.common.base.Verify.verifyNotNull
> com.google.common.base.Preconditions.checkNotNull
> org.apache.commons.lang.Validate.notNull
> ```
> 
> I think it would be good adding unit tests for these too

There is already a unit test for `com.google.common.base.Preconditions.checkNotNull` with [RV_RETURN_VALUE_IGNORED_Guava_Preconditions](https://github.com/spotbugs/spotbugs/blob/master/spotbugsTestCases/src/java/bugPatterns/RV_RETURN_VALUE_IGNORED_Guava_Preconditions.java) as sample class tested in [PreconditionsCheckNotNullCanIgnoreReturnValueTest](https://github.com/spotbugs/spotbugs/blob/master/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/PreconditionsCheckNotNullCanIgnoreReturnValueTest.java), but when running it, the `com.google.common.base.Preconditions` class needed for analysis was missing, and the test does not fail because of the missing class.

**Comment by gtoison** [comment]
@JuditKnoll would you know if the missing class error is because of the way we are running unit tests?
I was wondering if it could be a problem similar to https://github.com/spotbugs/spotbugs/issues/438

**Comment by JuditKnoll** [comment]
> @JuditKnoll would you know if the missing class error is because of the way we are running unit tests? I was wondering if it could be a problem similar to #438

Yes, it seems really similar.
The `AbstractIntegrationTest.performAnalysis()` adds the auxClassPathEntries from the `spotbugsTestCases/lib` directory, adding the necessary jars here, removes these messages. I think we should add the aux classpathes listed in the `spotbugsTestCases/build/spotbugs/auxclasspath/spotbugsMain` file, or if you have a better idea, I would love to hear it, but this should be in a separate PR.

P.S. I'm still looking into whether `com.google.common.base.Verify.verifyNotNull()`, `org.apache.commons.lang3.Validate.notNull()` are already handled, or just not found because of some connecting problem in the tests.

**Comment by JuditKnoll** [comment]
I addressed the issues mentioned in the reviews. However, when running the test separately from the IDE, the following is written to the std err*:
```
The following classes needed for analysis were missing:
  com.google.common.base.Verify
  org.apache.commons.lang3.Validate
  com.google.common.base.Preconditions
```
Locally I added these to the `spotbugsTestcases/lib`, which does solve this issue, and the tests in `IssueRequireNotNullTest` are still successful, but some other tests fail, which also fail on master if the necessary jars are added.
It looks like the functions are handled by the `IsNullValueDataflow`, and not by `callToAssertionMethod()` function.


*When running from terminal using gradle, these missing classes messages does not appear in the output, the build is successful. IMO if there are missing classes, the tests should fail.

**Review by JuditKnoll — APPROVED** [review]

**Review by gtoison — APPROVED** [review]
This LGTM, thank you!

**Review by ThrawnCA — APPROVED** [review]

**Comment by hazendaz** [comment]
fixes #651 and fixes #456

**Pull request merged**